In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re
import os

In [2]:
pc = pd.read_parquet(r'C:\Users\Abdon.RAMIREZ-BRICEN\OneDrive - Akkodis\Desktop\EDA_PC\Data\Raw\train.parquet',engine='fastparquet')

In [3]:
df = pc.copy()

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 912 entries, 0 to 911
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   laptop_ID         912 non-null    int64  
 1   Company           912 non-null    object 
 2   Product           912 non-null    object 
 3   TypeName          912 non-null    object 
 4   Inches            912 non-null    object 
 5   ScreenResolution  912 non-null    object 
 6   Cpu               912 non-null    object 
 7   Ram               912 non-null    object 
 8   Memory            912 non-null    object 
 9   Gpu               912 non-null    object 
 10  OpSys             912 non-null    object 
 11  Weight            912 non-null    object 
 12  Price_euros       912 non-null    float64
dtypes: float64(1), int64(1), object(11)
memory usage: 92.8+ KB


In [5]:
df.isnull().sum()

laptop_ID           0
Company             0
Product             0
TypeName            0
Inches              0
ScreenResolution    0
Cpu                 0
Ram                 0
Memory              0
Gpu                 0
OpSys               0
Weight              0
Price_euros         0
dtype: int64

In [6]:
df.head()

laptop_ID  Company                                  Product  \
0        599     Dell                            Inspiron 3179   
1         92     Asus  FX550IK-DM018T (FX-9830P/8GB/1TB/Radeon   
2        744       HP                                 ZBook 15   
3        729       HP                            Pavilion X360   
4        957  Toshiba                          Tecra A50-C-1ZV   

             TypeName Inches                ScreenResolution  \
0  2 in 1 Convertible   11.6            Touchscreen 1366x768   
1              Gaming   15.6               Full HD 1920x1080   
2         Workstation   15.6               Full HD 1920x1080   
3  2 in 1 Convertible   13.3  IPS Panel Touchscreen 1366x768   
4            Notebook   15.6                        1366x768   

                           Cpu   Ram     Memory                    Gpu  \
0     Intel Core M 7Y30 1.0GHz   4GB  128GB SSD  Intel HD Graphics 615   
1            AMD FX 9830P 3GHz   8GB    1TB HDD      AMD Radeon RX 560   
2  Intel Core i7 7700HQ 2.8GHz  16GB  256GB SSD    Nvidia Quadro M2200   
3   Intel Core i5 7200U 2.5GHz   8GB  256GB SSD  Intel HD Graphics 620   
4   Intel Core i5 6200U 2.3GHz   8GB  256GB SSD  Intel HD Graphics 520   

        OpSys  Weight  Price_euros  
0  Windows 10  1.39kg        603.0  
1  Windows 10  2.45kg        699.0  
2  Windows 10   2.6kg       2419.0  
3  Windows 10  1.58kg        836.6  
4  Windows 10   2.2kg       1119.0

In [7]:
df.tail()

laptop_ID Company                               Product  \
907        837    Acer                         Chromebook 11   
908        626    Acer  CB5-132T-C9KK (N3160/4GB/32GB/Chrome   
909        225    Dell                         Inspiron 3576   
910         17    Dell                         Inspiron 3567   
911         87    Dell                         Inspiron 7577   

               TypeName Inches                ScreenResolution  \
907             Netbook   11.6              IPS Panel 1366x768   
908  2 in 1 Convertible   11.6  IPS Panel Touchscreen 1366x768   
909            Notebook   15.6               Full HD 1920x1080   
910            Notebook   15.6               Full HD 1920x1080   
911              Gaming   15.6     IPS Panel Full HD 1920x1080   

                                      Cpu   Ram                Memory  \
907  Intel Celeron Dual Core N3060 1.6GHz   4GB    32GB Flash Storage   
908  Intel Celeron Quad Core N3160 1.6GHz   4GB    32GB Flash Storage   
909            Intel Core i5 8250U 1.6GHz   8GB             256GB SSD   
910            Intel Core i7 7500U 2.7GHz   8GB             256GB SSD   
911           Intel Core i7 7700HQ 2.8GHz  16GB  128GB SSD +  1TB HDD   

                            Gpu       OpSys  Weight  Price_euros  
907       Intel HD Graphics 400   Chrome OS  1.35kg       355.00  
908       Intel HD Graphics 400   Chrome OS  1.25kg       379.00  
909              AMD Radeon 520       Linux   2.2kg       677.35  
910          AMD Radeon R5 M430  Windows 10   2.2kg       745.00  
911  Nvidia GeForce GTX 1050 Ti  Windows 10  2.65kg      1349.00

In [8]:
df.columns

Index(['laptop_ID', 'Company', 'Product', 'TypeName', 'Inches',
       'ScreenResolution', 'Cpu', 'Ram', 'Memory', 'Gpu', 'OpSys', 'Weight',
       'Price_euros'],
      dtype='object')

In [9]:
df.shape

(912, 13)

In [10]:
df.describe()

laptop_ID  Price_euros
count   912.000000   912.000000
mean    663.118421  1111.262533
std     380.713558   682.109694
min       1.000000   174.000000
25%     339.750000   598.742500
50%     666.500000   977.000000
75%     985.250000  1472.650000
max    1320.000000  5499.000000

In [11]:
df[df.duplicated()]

Empty DataFrame
Columns: [laptop_ID, Company, Product, TypeName, Inches, ScreenResolution, Cpu, Ram, Memory, Gpu, OpSys, Weight, Price_euros]
Index: []

In [12]:
df_ob = df.select_dtypes('object')
for col in df_ob.columns:
    print(f'==========={col}===========')
    print(f'Numero de unicos: {df[col].nunique()}')
    print(f'Valores Unicos: {df[col].unique()}')
    print('\n')

===========Company===========
Numero de unicos: 19
Valores Unicos: ['Dell' 'Asus' 'HP' 'Toshiba' 'Lenovo' 'Google' 'LG' 'MSI' 'Acer'
 'Mediacom' 'Apple' 'Microsoft' 'Fujitsu' 'Huawei' 'Xiaomi' 'Vero' 'Chuwi'
 'Samsung' 'Razer']


===========Product===========
Numero de unicos: 484
Valores Unicos: ['Inspiron 3179' 'FX550IK-DM018T (FX-9830P/8GB/1TB/Radeon' 'ZBook 15'
 'Pavilion X360' 'Tecra A50-C-1ZV' 'Vostro 5468' '250 G6' 'Latitude 3580'
 'IdeaPad 310-15IKB' 'IdeaPad Y700-15ACZ' 'EliteBook 840' 'Inspiron 5567'
 'Pixelbook (Core' 'VivoBook E201NA' 'Latitude E5570' 'ProBook 650'
 'IdeaPad Y700-15ISK' 'Yoga 500-14ISK' 'Inspiron 5579' 'Vostro 3559'
 'Gram 15Z970' 'Yoga 11e' 'XPS 13' 'Latitude 5480' 'ProBook 450'
 'GP62M Leopard' 'Inspiron 3567' 'Inspiron 5770' 'ThinkPad Yoga'
 'Thinkpad P51' 'Latitude 7280' 'Legion Y720-15IKB' 'Precision M5520'
 'Aspire 3' 'Omen 17-an012dx' 'VivoBook Max' 'E5 774G' 'Zbook 15'
 'Inspiron 3552' 'B51-80 (i5-6200U/8GB/1TB/Radeon' 'SmartBook Edge'
 'IdeaPad 320

In [13]:
df['Inches'] = df['Inches'].astype(float)

In [14]:
df

laptop_ID  Company                                  Product  \
0          599     Dell                            Inspiron 3179   
1           92     Asus  FX550IK-DM018T (FX-9830P/8GB/1TB/Radeon   
2          744       HP                                 ZBook 15   
3          729       HP                            Pavilion X360   
4          957  Toshiba                          Tecra A50-C-1ZV   
..         ...      ...                                      ...   
907        837     Acer                            Chromebook 11   
908        626     Acer     CB5-132T-C9KK (N3160/4GB/32GB/Chrome   
909        225     Dell                            Inspiron 3576   
910         17     Dell                            Inspiron 3567   
911         87     Dell                            Inspiron 7577   

               TypeName  Inches                ScreenResolution  \
0    2 in 1 Convertible    11.6            Touchscreen 1366x768   
1                Gaming    15.6               Full HD 1920x1080   
2           Workstation    15.6               Full HD 1920x1080   
3    2 in 1 Convertible    13.3  IPS Panel Touchscreen 1366x768   
4              Notebook    15.6                        1366x768   
..                  ...     ...                             ...   
907             Netbook    11.6              IPS Panel 1366x768   
908  2 in 1 Convertible    11.6  IPS Panel Touchscreen 1366x768   
909            Notebook    15.6               Full HD 1920x1080   
910            Notebook    15.6               Full HD 1920x1080   
911              Gaming    15.6     IPS Panel Full HD 1920x1080   

                                      Cpu   Ram                Memory  \
0                Intel Core M 7Y30 1.0GHz   4GB             128GB SSD   
1                       AMD FX 9830P 3GHz   8GB               1TB HDD   
2             Intel Core i7 7700HQ 2.8GHz  16GB             256GB SSD   
3              Intel Core i5 7200U 2.5GHz   8GB             256GB SSD   
4              Intel Core i5 6200U 2.3GHz   8GB             256GB SSD   
..                                    ...   ...                   ...   
907  Intel Celeron Dual Core N3060 1.6GHz   4GB    32GB Flash Storage   
908  Intel Celeron Quad Core N3160 1.6GHz   4GB    32GB Flash Storage   
909            Intel Core i5 8250U 1.6GHz   8GB             256GB SSD   
910            Intel Core i7 7500U 2.7GHz   8GB             256GB SSD   
911           Intel Core i7 7700HQ 2.8GHz  16GB  128GB SSD +  1TB HDD   

                            Gpu       OpSys  Weight  Price_euros  
0         Intel HD Graphics 615  Windows 10  1.39kg       603.00  
1             AMD Radeon RX 560  Windows 10  2.45kg       699.00  
2           Nvidia Quadro M2200  Windows 10   2.6kg      2419.00  
3         Intel HD Graphics 620  Windows 10  1.58kg       836.60  
4         Intel HD Graphics 520  Windows 10   2.2kg      1119.00  
..                          ...         ...     ...          ...  
907       Intel HD Graphics 400   Chrome OS  1.35kg       355.00  
908       Intel HD Graphics 400   Chrome OS  1.25kg       379.00  
909              AMD Radeon 520       Linux   2.2kg       677.35  
910          AMD Radeon R5 M430  Windows 10   2.2kg       745.00  
911  Nvidia GeForce GTX 1050 Ti  Windows 10  2.65kg      1349.00  

[912 rows x 13 columns]

In [15]:
df['Ram(GB)'] = df['Ram'].replace('GB','',regex=True)

### Comenzamos con la limpieza de datos de la columna Product.

In [16]:
df['Product'] = df['Product'].str.replace(r'[()\-/]', ' ', regex=True)
 # quitamos caracteres especiales

In [17]:
df['Product'] = df.apply(lambda row: ' '.join([palabra for palabra in row['Product'].split() if palabra not in row['Cpu']]), axis=1)
# Con este lambda quitamos todo lo que esta en la columna CPU de la columna Product

In [18]:
sig = r'\b(\d+GB|\d+TB|FX-\d+|i[3579]-\d{4}\w*|Celeron|Radeon|GeForce|Intel HD|Linux|Chrome|W10|W7|FHD|No|12.3)\b'
df['Product'] = df['Product'].str.replace(sig, '', regex=True).str.strip()
# Quitamos las memorias, procesadores y algunos sistemas operativos

In [19]:
patron = r'\d+(\.\d+)?\"'
df['Product'] = df['Product'].apply(lambda x: re.sub(patron,'',x))
# Quitamos del producto las inchas ya que las tenemos en una columna aparte.

En esta columna no dividimos nada, solo quitamos datos que ya teniamos en otras columnas para simplificarlos.

### Pasamos a la columna TypeName

In [20]:
df['TypeName'] = df['TypeName'].apply(lambda x: 'Convertible' if x == '2 in 1 Convertible' else x )

# Siguiendo el patron de los otros TypeName elinimanos el "2 in 1" del TypeName, para solo usar 1 palabra.

### Columna ScreenResolution:

In [21]:
df['Resolution'] = df['ScreenResolution'].str.extract(r'(\d+x\d+)')
# Separamos la Resolucion.

df['PanelType'] = df['ScreenResolution'].str.findall(r'(IPS Panel|Retina Display|Quad HD\+|Full HD|4K Ultra HD)').str.join(', ')
df['PanelType'] = df['PanelType'].apply(lambda x: 'No penal type' if x == "" else x)
# separamos el tipo de panel

df['Touchscreen'] = df['ScreenResolution'].str.contains(r'Touchscreen', case=False, na=False)
df['Touchscreen'] = df['Touchscreen'].apply(lambda x: 1 if x else 0)
# Transformamos a tipo bool si es touch o no.


In [22]:
df = df[['laptop_ID','Company','Product','TypeName','Inches','Resolution','PanelType','Touchscreen','Cpu'
         ,'Ram(GB)','Memory','Gpu','OpSys','Weight']]
# Ordenamos el DataFrame


Esta columna la separamos en 3: Resolution, PanelType y TouchScreen. El ultimo booleano.

### Columna CPU

In [23]:
pat = r'\b(\d+(?:\.\d+)?)\s*GHz\b'
df['ProcessorSpeed(GHz)'] = df['Cpu'].str.extract(f'{(pat)}')

# Primero extraemos la velocidad del procesador.

C:\Users\Abdon.RAMIREZ-BRICEN\AppData\Local\Temp\ipykernel_39996\1311523093.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['ProcessorSpeed(GHz)'] = df['Cpu'].str.extract(f'{(pat)}')


In [24]:
df['Cpu'] = df.apply(lambda row: ' '.join([palabra for palabra in row['Cpu'].split() if palabra not in row['ProcessorSpeed(GHz)']]),axis=1)
df['ProcessorSpeed(GHz)'] = df['ProcessorSpeed(GHz)'].astype(float)
# Eliminamos la velocidad de la columna Cpu

C:\Users\Abdon.RAMIREZ-BRICEN\AppData\Local\Temp\ipykernel_39996\3472316383.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Cpu'] = df.apply(lambda row: ' '.join([palabra for palabra in row['Cpu'].split() if palabra not in row['ProcessorSpeed(GHz)']]),axis=1)
C:\Users\Abdon.RAMIREZ-BRICEN\AppData\Local\Temp\ipykernel_39996\3472316383.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['ProcessorSpeed(GHz)'] = df['ProcessorSpeed(GHz)'].astype(float)


In [25]:
# Extraemos el tipo de procesador (i3, i5, i7, etc.)
df['ProcessorType'] = df['Cpu'].str.extract(r'(i3|i5|i7|i9|AMD Ryzen|AMD A4-Series|AMD A9-Series|AMD E-Series|AMD A6-Series|AMD|Xeon|FX|Celeron|Pentium|Atom|Cortex|M)')

C:\Users\Abdon.RAMIREZ-BRICEN\AppData\Local\Temp\ipykernel_39996\96076574.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['ProcessorType'] = df['Cpu'].str.extract(r'(i3|i5|i7|i9|AMD Ryzen|AMD A4-Series|AMD A9-Series|AMD E-Series|AMD A6-Series|AMD|Xeon|FX|Celeron|Pentium|Atom|Cortex|M)')


In [26]:
df['Core'] = df['Cpu'].str.extract(r'\b(Dual|Quad|Core)')
df['Core'] = df['Core'].map({'Core':0,'Dual':2,'Quad':4})
df['Core'] = df['Core'].fillna(0) 
# Extraemos los nucleos de la columna Cpu, y les asignamos un numero.

In [27]:
df['Cpu'] = df['Cpu'].replace(pat,'',regex=True)

In [28]:
def obtener_identificador(cpu):
    if not isinstance(cpu, str):
        return None
    palabras = cpu.split()
    if not palabras:
        return None
    if len(palabras[-1]) > 2:
        return palabras[-1]
    elif len(palabras) >= 2:
        return palabras[-2]
    else:
        return None
    
    # Descubrimos que la mayoria de veces, la palabra final es el modelo de identificacion. Creamos esta funcion para extraerla.

In [29]:
df['ModelIdentifier'] = df['Cpu'].apply(obtener_identificador) # Aplicamos la funcion
df['ModelIdentifier'] = df['ModelIdentifier'].apply(lambda x: '7200' if x == 'M' or x == 'Core' else x)

# Quitamos 'M' y 'Core' que no corresponden a modelos de identificacion.

La columna Cpu la dividimos en 4: ProcessorSpeed, ProcessorType, ModelIdentifier y Core. Core le pusimos valor a los que especificaban el nuemro de nucleos, ejemplo: los que solo tienen 'core', no especifica esos pusimos 0, dual con 2 y Quad 4.

In [30]:
df = df[['laptop_ID', 'Company', 'Product', 'TypeName', 'Inches', 'Resolution',
       'PanelType', 'Touchscreen', 'ProcessorType','ModelIdentifier','Core','ProcessorSpeed(GHz)', 'Ram(GB)', 'Memory', 'Gpu', 'OpSys',
       'Weight']]

# Ordenamos el DataFrame

### Columna Memory

In [31]:
df['Memory SSD GB'] = df['Memory'].apply(lambda x: x if 'SSD' in x else 0).str.split('+').str[0].fillna(0)
df['Memory HDD GB'] = df['Memory'].apply(lambda x: x if 'HDD' in x else 0).str.split('+').str[-1].fillna(0)
df['Memory FL STO GB'] = df['Memory'].apply(lambda x: x if 'Flash' in x else 0).str.split('+').str[0].fillna(0)
df['Memory Hybrid GB'] = df['Memory'].apply(lambda x: x if 'Hybrid' in x else 0)
# Separamos las memorias por tipo

In [32]:
df['Memory SSD GB'] = df['Memory SSD GB'].str.replace(r'[a-zA-Z]','',regex=True).str.strip().astype(float).fillna(0)
df['Memory SSD GB'] = df['Memory SSD GB'].apply(lambda x: x*1000 if x <= 2 else x)
df['Memory HDD GB'] = df['Memory HDD GB'].str.replace(r'[a-zA-Z]','',regex=True).str.strip().astype(float).fillna(0)
df['Memory HDD GB'] = df['Memory HDD GB'].apply(lambda x: x*1000 if x <= 2 else x)
df['Memory FL STO GB'] = df['Memory FL STO GB'].str.replace(r'[a-zA-Z]','',regex=True).str.strip().astype(float).fillna(0)
df['Memory FL STO GB'] = df['Memory FL STO GB'].apply(lambda x: x*1000 if x <= 2 else x)
df['Memory Hybrid GB'] = df['Memory Hybrid GB'].str.replace(r'[a-zA-Z]','',regex=True).str.strip().astype(float).fillna(0)
df['Memory Hybrid GB'] = df['Memory Hybrid GB'].apply(lambda x: x*1000 if x <= 2 else x)

# Eliminamos todas las letras para dejar solo dato numerico en las columnas.

In [33]:
df = df[['laptop_ID', 'Company', 'Product', 'TypeName', 'Inches', 'Resolution',
       'PanelType', 'Touchscreen', 'ProcessorType','ModelIdentifier','Core','ProcessorSpeed(GHz)', 'Ram(GB)', 'Memory SSD GB','Memory HDD GB','Memory FL STO GB','Memory Hybrid GB', 'Gpu', 'OpSys',
       'Weight']]

La columna Memory la dividimos en 2: Memory 1 y Memory 2. Los pc que no tenian Memory 2 los llenamos con "No Memory 2"

### Columna Gpu

In [34]:
df.loc[:,'Gpu'] = df['Gpu'].str.replace(r'^(Intel|AMD)\s*', '', case=False, regex=True).str.strip()
# Quitamos Intel y AMD de la columna Gpu debido a que este dato lo tenemos en otra columna.

df[['GpuBrand', 'GpuModel']] = df['Gpu'].str.extract(r'([A-Za-z]+(?:\s+[A-Za-z]+)*)\s*([\d\w]+)?')
# Separamos entre marca y modelo la columna GPU.
df['GpuModel'] = df['GpuModel'].fillna('0')

In [35]:
df['GpuModel'].unique()

array(['615', '560', '2200', '620', '520', '920MX', '9', '7', '500',
       '960M', '5', '1050', '530', '2200M', '0', '1060', '1200', '580',
       '940MX', '4', '400', '640', '1070', '630', '930MX', '405', '550',
       '150', '1050Ti', '130', '2', '5130M', '1000M', '515', '17M',
       '1050M', '3', '980M', '960', '6150M', '505', '930M', '965M', '980',
       '950M', '1080', '650', '150MX', '540', '940M', '920M', '2000M',
       '520M', '6000', '510', '5300', '860', '4190M', '970M', '920'],
      dtype=object)

In [36]:
df = df[['laptop_ID', 'Company', 'Product', 'TypeName', 'Inches', 'Resolution',
       'PanelType', 'Touchscreen', 'ProcessorType','ModelIdentifier','Core','ProcessorSpeed(GHz)', 
       'Ram(GB)', 'Memory SSD GB','Memory HDD GB','Memory FL STO GB','Memory Hybrid GB', 'GpuBrand','GpuModel', 'OpSys',
       'Weight']]

La columna Gpu la dividimos en 2: GpuBrand y GpuModel. Marca y modelo repsectivamente.

### Columna OpSys

In [37]:
df['OpSys'] = df['OpSys'].str.replace(r'(10|10\s*S|7|X)\s*$', '', regex=True).str.strip()
df['OpSys'] = df['OpSys'].str.replace('macOS','Mac OS')

In [38]:
df['Weight(KG)'] = df['Weight'].replace('kg','',regex=True)
df['Weight(KG)'] = df['Weight(KG)'].astype(float)

# Le quitamos kg a la columna Weight para un mejor analisis.

In [39]:
df = df[['laptop_ID', 'Company', 'Product', 'TypeName', 'Inches', 'Resolution',
       'PanelType', 'Touchscreen', 'ProcessorType','ModelIdentifier','Core','ProcessorSpeed(GHz)', 
       'Ram(GB)', 'Memory SSD GB','Memory HDD GB','Memory FL STO GB','Memory Hybrid GB', 'GpuBrand','GpuModel', 'OpSys',
       'Weight(KG)']]

Simplemente borramos Las versiones, creemos que el analisis seria mejor asi.

In [40]:
df['Ram(GB)'] = df['Ram(GB)'].astype(int)

In [41]:
df['TypeName'] = df['TypeName'].map({'Convertible':1,'Gaming':2, 'Workstation':3, 'Notebook':4, 'Ultrabook':5, 'Netbook':6})

In [42]:
df[['Ancho', 'Alto']] = df['Resolution'].str.split('x', expand=True).astype(int)
df['Pixeles'] = df['Ancho'] * df['Alto']
df['AspectRatio'] = df['Ancho'] / df['Alto']

In [43]:
df['ProcessorType'] = df['ProcessorType'].map({'i7':1,'Atom':2, 'AMD E-Series':3, 'i3':4, 'i5':5, 'Celeron':6,
                                          'Pentium':7,'AMD A9-Series':8,'M':9,'AMD':10, 'Xeon':11, 'AMD A6-Series':12, 'Cortex':13, 'AMD Ryzen':4})

In [44]:
df['ModelIdentifier'] = [re.sub(r'[A-Za-z&-]', '', modelo) for modelo in df['ModelIdentifier']]
df['ModelIdentifier'] = df['ModelIdentifier'].astype(int)

In [45]:
df['GpuModel'] = [re.sub(r'[A-Za-z&-]', '', modelo) for modelo in df['GpuModel']]
df['GpuModel'] = df['GpuModel'].astype(int)

In [46]:
df = df.drop(columns=['Product','Resolution'])

In [47]:
from sklearn.preprocessing import OneHotEncoder,LabelEncoder
encoder = OneHotEncoder(handle_unknown='ignore',sparse_output=False)
df_encoded = encoder.fit_transform(df[['Company']])

# Convertir a DataFrame
df_encoded = pd.DataFrame(df_encoded, columns=encoder.get_feature_names_out(['Company']))
df = pd.concat([df,df_encoded],axis=1)
df = df.drop(columns='Company')

In [48]:
encoder_panel = OneHotEncoder(handle_unknown='ignore',sparse_output=False)
df_encoded_panel = encoder_panel.fit_transform(df[['PanelType']])

# Convertir a DataFrame
df_encoded_panel = pd.DataFrame(df_encoded_panel, columns=encoder_panel.get_feature_names_out(['PanelType']))
df = pd.concat([df,df_encoded_panel],axis=1)
df = df.drop(columns='PanelType')

In [49]:
encoder_op = OneHotEncoder(handle_unknown='ignore',sparse_output=False)
df_encoded_op = encoder_op.fit_transform(df[['OpSys']])

# Convertir a DataFrame
df_encoded_op = pd.DataFrame(df_encoded_op, columns=encoder_op.get_feature_names_out(['OpSys']))
df = pd.concat([df,df_encoded_op],axis=1)
df = df.drop(columns='OpSys')

In [50]:
enc = LabelEncoder()
df['GpuBrand'] = enc.fit_transform(df['GpuBrand'])


In [51]:
df.to_csv('train_clean.csv',index = 'laptop_ID')

## Terminando con la limpieza y future engineer, nos quedamos con estas variables.

|Variable|Descripcion|
|----|----|
|Laptop_ID|Indice.|
|Company|Compañia fabricante.|
|Product|Nombre del PC.|
|TypeName|Tipo de PC.|
|Inches|Tamaño de pantalla.|
|Resolution|Resolucion de pantalla.|
|PanelType|Tipo de panel.|
|TouchScreen|Tipo Bool, si la pantalla es touch o no.|
|ProcessorType|Tipo de procesador.|
|ModelIdentifier|Identificador del modelo.|
|Core|Cantidad de Nucleos.|
|ProcessorSpeed|Velocidad de Procesador.|
|Ram|Memoria Ram.|
|Memory SSD GB|Memoria SSD|
|Memory HDD GB|Memoria Nro SSD|
|Memory FL STO GB|Memoria Flash Storage|
|Memory Hybrid GB|MEmoria Hibrida|
|GpuBrand|Marca de Gpu.|
|GpuModel|Modelo de Gpu.|
|OpSys|Sistema operativo.|
|Weight|Peso del PC.|
|Price_euros|Precio de PC.|


## Analisis Univariante.

In [82]:
import sys
sys.path.append(r'C:\Users\Abdon.RAMIREZ-BRICEN\OneDrive - Akkodis\Documents\Analisis_datos_eda\scripts')

In [83]:
# os.chdir(r'C:\Users\Abdon.RAMIREZ-BRICEN\OneDrive - Akkodis\Documents\Analisis_datos_eda\scripts\ClaseAbdon.py')

In [84]:
from ClaseAbdon import DataFrameAnalyzer
# Importo mi clase

In [85]:
clase = DataFrameAnalyzer(df)

In [ ]:
clase.test_Ksmirnov()

In [87]:
from scipy.stats import shapiro, anderson, normaltest

In [ ]:
num = df.select_dtypes(include='number')

for n in num:
    stat, p = normaltest(df[n])
    print(f'{n}: Dagostino test         statistic: {stat}, p-value: {p}')

In [ ]:
clase.resumen()

Podemos observar que hay una alta cardinalidad en el Nombre de producto y en el precio.
En el resto de variables hay una cardinalidad baja, es normal porque muchos PC tienen especificaciones "parecidas".

In [ ]:
clase.describe_numeric()

Poco que mirar en las variables numericas, touchscreen es tipo bool y core "Numerica discreta". 
Con respecto al precio vemos que la mediana y la media estan relativamente "cerca", la moda es 499 y tiene una STD un poco elevada. lo que los precios varian considerablemente.

In [ ]:
clase.describe_categorical()

En este DF vemos que la compañia Dell es la que mas frecuenta entre las marcas. El producto "XPS 13" el mas frecuente entre ellos, el sistema operativo mas frecuente es Window.. etc.

In [ ]:
clase.plot_numeric()

In [184]:
# Q1_price = df['Price_euros'].quantile(0.25)
# Q3_price = df['Price_euros'].quantile(0.75)
# IQR_price = Q3_price - Q1_price

# upper_limit_price = Q3_price + 1.5 * IQR_price

# # Identificar valores fuera del rango IQR
# outl_price = df[df['Price_euros'] > upper_limit_price]
# outl_price[['Company','Product','Price_euros']]

En la variable de precios vemos que hay ciertos outliers, como vimos anteriormente el max superaba los 5000€. Es normal que existan PC mas costosos que otros.

In [ ]:
Q1_weight = df['Weight(KG)'].quantile(0.25)
Q3_weight = df['Weight(KG)'].quantile(0.75)
IQR_weight = Q3_weight - Q1_weight

upper_limit_weight = Q3_weight + 1.5 * IQR_weight

# Identificar valores fuera del rango IQR
outl_weight = df[df['Weight(KG)'] > upper_limit_weight]
outl_weight[['Company','Product','Weight(KG)']]

En la columna weigh tenemos estos outliers, Alienware 17 es el pc que mas se repite en los outliers.

In [ ]:
df_numeric = df.select_dtypes('number')
correlation_matrix = df_numeric.corr()
plt.figure(figsize=(10, 8))
sns.heatmap(correlation_matrix, annot=True, cmap='coolwarm', fmt=".2f", linewidths=0.5)
plt.title("Mapa de Correlación")
plt.show()

- Vemos que en el precio tiene mas corr con la velocidad de procesador y la 2DA memoria.
- El peso y la 2DA memoria tambien tiene un relacion relativamente "alta".
- El peso y las inchas estan muy correlaciones, mientras mas inchas mas peso.


In [ ]:
company_counts = df['Company'].value_counts()
plt.figure(figsize=(12, 6))
sns.barplot(x=company_counts.index, y=company_counts.values)
plt.title('Cantidad de productos por compañía')
plt.xlabel('Compañía')
plt.ylabel('Cantidad de productos')
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.show()

Observamos que Dell, Lenovo y HP dominan entre las demas compañias.

In [ ]:
type_counts = df['TypeName'].value_counts()
plt.figure(figsize=(12, 6))
sns.barplot(x=type_counts.index, y=type_counts.values)
plt.title('Tipos de PC')
plt.xlabel('Tipos de PC')
plt.ylabel('Cantidad de tipos')
plt.tight_layout()
plt.show()

Vemos que los tipo Notebook son los que mayormente hay.

In [ ]:
touch_counts = df['Touchscreen'].value_counts()
plt.figure(figsize=(12, 6))
sns.barplot(x=touch_counts.index, y=touch_counts.values)
plt.title('Cantidad de touchscreen')
plt.xlabel('No/Yes')
plt.ylabel('Cantidad de PC')
plt.tight_layout()
plt.show()

Vemos que hay gran diferencia entre los PC que no tienen pantalla tactil y los que si.

In [ ]:
plt.figure(figsize=(12, 6))
sns.histplot(df['Resolution'], discrete=True, kde=False)
plt.title('Resoluciones')
plt.ylabel('Cantidad de resolucion PC')
plt.xticks(rotation=45, ha='right')
plt.show()

La resolucion 1920X1080, es la mas comun entre estos pc.

In [ ]:
plt.figure(figsize=(12, 6))
sns.histplot(df['PanelType'], discrete=True, kde=False)
plt.title('Tipo de Panel')
plt.ylabel('Cantidad de tipo de paneles.')
plt.xticks(rotation=45, ha='right')

plt.show()

El panel Full HD, el ,as comun entre los pc.

In [ ]:
proce_counts = df['ProcessorType'].value_counts()
plt.figure(figsize=(12, 6))
sns.barplot(x=proce_counts.index, y=proce_counts.values)
plt.title('Cantidad por procesadores')
plt.xlabel('Procesadores')
plt.ylabel('Cantidad de PC')
plt.xticks(rotation=45, ha='right')
# plt.tight_layout()
plt.show()

In [ ]:
core = df['Core'].value_counts()
core.index = core.index.map({0: 'Sin identificar', 2: 'Dual', 4: 'Quad'})

plt.figure(figsize=(8, 6))
core.plot(kind='bar', color='skyblue', edgecolor='black')

plt.title('Distribucion de Core', fontsize=14)
plt.xlabel('Tipo de Core', fontsize=12)
plt.ylabel('Cantidad', fontsize=12)
plt.xticks(rotation=0)
plt.show()

Solo tenemos unos pocos Cores especificados, nos pueden servir en un futuro para poder hacer comparaciones entre laptops.

In [ ]:
ram_counts = df['Ram(GB)'].value_counts()
plt.figure(figsize=(12, 6))
sns.barplot(x=ram_counts.index, y=ram_counts.values)
plt.title('Cantidad de Memorias Ram')
plt.xlabel('RAM')
plt.ylabel('Cantidad de PC')
plt.tight_layout()
plt.show()

Para la sorpresa de nadia.. La de 8GB es la mas frecuente entre los 912 pc.

In [ ]:
op_counts = df['OpSys'].value_counts(ascending=True)
plt.figure(figsize=(12, 6))
plt.plot(
    op_counts.values, 
    op_counts.index, 
    'o-', color='skyblue', markersize=10, markeredgecolor='black'
)

plt.title('Cantidad de Sistemas Operativos', fontsize=14)
plt.xlabel('Cantidad de PC', fontsize=12)
plt.ylabel('Sistema Operativo', fontsize=12)
plt.grid(axis='x', linestyle='--', alpha=0.7)
plt.tight_layout()
plt.show()

Terminamos con que WINDOWS es el sistema operativo mas comun entre todos los pc.